python/ray/ml/checkpoint.py

```diff
@@ -332,12 +330,36 @@ def to_directory(self, path: Optional[str] = None, dedup: bool = True) -> str:
                     shutil.copytree(local_path, path)
             elif external_path:
                 # If this exists on external storage (e.g. cloud), download
-                download_from_uri(uri=external_path, local_path=path)
+                download_from_uri(uri=external_path, local_path=path, filelock=False)
             else:
                 raise RuntimeError(
                     f"No valid location found for checkpoint {self}: {self._uri}"
                 )
 
+    def to_directory(self, path: Optional[str] = None) -> str:
+        """Write checkpoint data to directory.
+
+        Args:
+            path: Target directory to restore data in. If not specified,
+                will create a temporary directory.
+
+        Returns:
+            str: Directory containing checkpoint data.
+        """
+        path = path if path is not None else self._get_temporary_checkpoint_dir()
+
+        _make_dir(path)
+
+        try:
+            with FileLock(f"{path}.lock", timeout=0):
+                self._to_directory(path)
+        except TimeoutError:
+            # if the directory is already locked, then wait but do not do anything.
+            with FileLock(f"{path}.lock", timeout=-1):
+                pass
+            if not os.path.exists(path):
```

Comment:
Can we raise an error instead of recursively calling this method? It seems this shouldn't happen.

Reply:
👍

Reply:
I see. How about creating the delete ref file first then? So then if you create the del lock and then acquire the file lock, that eliminates the race condition.

Reply:
Yep, that should work